python.md

```diff
@@ -1,103 +1,408 @@
+dict.values()
+"key" in dict    # let's say this returns False, then...
+dict["key"]      # ...this raises KeyError
+dict.get("key")  # ...this returns None
+dict.setdefault("key", 1)
+```
 ### Casting
 
-    int(str)
-    float(str)
-    str(int)
-    str(float)
-    'string'.encode()
+``` python
+int(str)
+float(str)
+str(int)
+str(float)
+'string'.encode()
+ord('A') # 65
+chr(65)  # A
+```
+
+### Iteration
+
+#### For Loops
+``` python
+for item in ["a", "b", "c"]:
+for i in range(4):        # 0 to 3
+for i in range(4, 8):     # 4 to 7
+for i in range(1, 9, 2):  # 1, 3, 5, 7
+for key, val in dict.items():
+for index, item in enumerate(list):
+```
+
+#### While Loops
+``` python
+current_value = 1
+while current_value <= 5:
+    print(current_value)
+    current_value += 1
+```
 
 ### Comprehensions
 
-    [fn(i) for i in list]            # .map
-    map(fn, list)                    # .map, returns iterator
-    
-    filter(fn, list)                 # .filter, returns iterator
-    [fn(i) for i in list if i > 0]   # .filter.map
+``` python
+[fn(i) for i in list]                   # .map
+map(fn, list)                           # .map, returns iterator
+map(lambda x, y: x+y, [1,2,3], [3,2,1]) # iterator: [4,4,4]
 
-### Regex
+filter(fn, list)                        # .filter, returns iterator
+list(filter(lambda x: x%2==1, [1,2,3])) # [1,3]
+
+[fn(i) for i in list if i > 0]          # .filter.map
+```
+
+### If Statements
+
+``` python
+if site == 'python.org':
+    print('python standard library documentation')
+else if site.startswith('devhints') and site.endswith('.io'):
+    print('TL;DR for developer documentation - a ridiculous collection of cheatsheets')
+else if not site.startswith('google'):
+    print('non-Google website')
+else:
+    pass
+```
+
+### User Input
+
+``` python
+name = input("What's your name? ")
+print(f"Hello, {name}!")
+```
+
+### Functions
+
+#### A Simple Function
+``` python
+def greet_user():
+    """Display a simple greeting."""
+    print("Hello!")
+greet_user()
+```
+
+#### Passing an Argument
+
+``` python
+def greet_user(username):
+    """Display a personalized greeting."""
+    print("Hello, " + username + "!")
+greet_user('jesse')
+```
+
+#### Default Values for parameters
+
+``` python
+def make_pizza(topping='bacon'):
+    """Make a single-topping pizza."""
+    print("Have a " + topping + " pizza!")
+make_pizza()
+make_pizza('pepperoni')
+```
+#### Returning a Value
+``` python
+def add_numbers(x, y):
+    """Add two numbers and return the sum."""
+    return x + y
+sum = add_numbers(3, 5)
+print(sum)
+```
+
+#### Lambda
+``` python
+add_numbers = lambda x, y: x+y
+sum = add_numbers(3, 5)
+print(sum)
+```
+
+### Classes
+
+#### Class Creation
+``` python
+class Dog():
+    """Represent a dog."""
+    def __init__(self, name):
+        """Initialize dog object."""
+        self.name = name
+        
+    def sit(self):
+        """Simulate sitting."""
+        print(self.name + " is sitting.")
```

Comment:
```suggestion
        print(self.name, "is sitting.")
```
Could be nice to show that print takes multiple parameters and so doesn’t require string concat like other languages